fix: fee calculation formula (#1092)
starknet-io/starknet-docs: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/fee-mechanism.adoc

The following formula describes the overall fee, stem:[F], for a transaction:

// OLD
// [stem]
// ++++
// F = \text{gas_price}\cdot\left(\max_k v_k w_k + 0.9\cdot\text{calldata_cost}\cdot\left(2(n+m) + 3t + \sum\limits_{i=1}^t q_i +\ell\right)\right)
// ++++

[stem]
++++
\begin{align}
F  = \text{gas_price}\cdot&\Bigg(\max_k v_k w_k + \\
& + \; \text{da_calldata_cost}\left(2n+2(m-1) + \ell + 3t + \sum\limits_{i=1}^t q_i\right)\\ 
& - \; \text{contract_update_discount}\cdot n - 240 \\
& + \; \text{message_calldata_cost}\cdot\left(3t + \sum\limits_{i=1}^t q_i\right) \\
& + \; \text{storage_write_cost}\cdot t\Bigg)
\end{align}
++++

where:

* stem:[$v$] is a vector that represents resource usage, where each of its entries, stem:[$v_k$], corresponds to different resource types: Cairo steps and number of applications of each builtin.
+
For more information see xref:#calculation_of_computation_costs[Calculation of computation costs].
* stem:[$n$] is xref:#storage_updates[the number of unique contracts updated]. Note that this also takes into account contract deployment, even if the storage of the newly deployed contract was untouched, i.e. stem:[$n\ge\ell$].

> * stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Note that stem:[$m\ge 1$] always holds since the sequencer balance is always updated (this is not charged for).

Reviewer: ```suggestion
* stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Notice that stem:[$m\ge 1$] is always true, because the sequencer's balance is always updated, which does not incur any fee.
```

* stem:[$t$] is the number of L2->L1 messages sent, where the corresponding payload sizes are denoted by stem:[$q_1,...,q_t$].

> * stem:[$\ell$] is the number of contracts whose class was changed. This can happen on deployment or on applying the `replace_class` syscall.

Reviewer: ```suggestion
* stem:[$\ell$] is the number of contracts whose class was changed, which can happen on deployment of a new version of the class, or on applying the `replace_class` syscall.
```

* stem:[$w$] is the xref:#calculation_of_computation_costs[`CairoResourceFeeWeights`] vector.

> * stem:[$\text{da_calldata_cost}$] is 551 gas per 32-byte word. This cost is derived as follows: 512 gas per 32-byte word for calldata, ~100gas for on-chain hashing that happens for every sent word, and 10% discount because the sequencer does not incur additional costs for repeated updates to the same storage slot within a single block, resulting in 551gas.

Reviewer: ```suggestion
* stem:[$\text{da_calldata_cost}$] is 551 gas per 32-byte word. This cost is derived as follows: 
+
** 512 gas per 32-byte word for calldata.
** ~100 gas for onchain hashing that happens for every sent word.
** a 10% discount, because the sequencer does not incur additional costs for repeated updates to the same storage slot within a single block.
```

This section shows the formula for determining a transaction's fee. The following sections describe how this formula was derived.

The following formula describes the overall fee, stem:[F], for a transaction:

// OLD
// [stem]
// ++++
// F = \text{gas_price}\cdot\left(\max_k v_k w_k + 0.9\cdot\text{calldata_cost}\cdot\left(2(n+m) + 3t + \sum\limits_{i=1}^t q_i +\ell\right)\right)
// ++++

[stem]
++++
\begin{align}
F  = \text{gas_price}\cdot&\Bigg(\max_k v_k w_k + \\
& + \; \text{da_calldata_cost}\left(2n+2(m-1) + \ell + 3t + \sum\limits_{i=1}^t q_i\right)\\ 
& - \; \text{contract_update_discount}\cdot n - 240 \\
& + \; \text{message_calldata_cost}\cdot\left(3t + \sum\limits_{i=1}^t q_i\right) \\
& + \; \text{storage_write_cost}\cdot t\Bigg)
\end{align}
++++

where:

* stem:[$v$] is a vector that represents resource usage, where each of its entries, stem:[$v_k$], corresponds to different resource types: Cairo steps and number of applications of each builtin.
+
For more information see xref:#calculation_of_computation_costs[Calculation of computation costs].

> * stem:[$n$] is xref:#storage_updates[the number of unique contracts updated]. Note that this also takes into account contract deployment, even if the storage of the newly deployed contract was untouched, i.e. stem:[$n\ge\ell$].

Reviewer: ```suggestion
* stem:[$n$] is xref:#storage_updates[the number of unique contracts updated], which includes changes to, or new deployments of, classes of existing contract instances, even if the storage of the newly deployed contract is untouched. In other words, stem:[$n\ge\ell$].
```

[stem]
++++
F = \text{gas_price}\cdot\left(\max_k v_k w_k + \text{calldata_cost}\cdot\left(2n+2(m-2) + 3t + \sum\limits_{i=1}^t q_i +3\ell\right) - \text{contract_update_discount}\cdot n + 310 + \text{storage_write_cost}\cdot t\right)
++++

where:

> * stem:[$v$] is a vector that represents resource usage, where each of its entries, stem:[$v_k$], corresponds to different resource types: Cairo steps and number of applications of each builtin.

Reviewer: I asked two folks to take a look at this, and one of them didn't understand what k is. The other person explained that _k_ is a number that corresponds to one of the resources listed in Table 1, so k is a value 1-8, (or 0-7?). Which means that the table should have a column for resource identifier _k_.

+
For more information see xref:#calculation_of_computation_costs[Calculation of computation costs].
* stem:[$n$] is xref:#storage_updates[the number of unique contracts updated].

> * stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Note that stem:[$m\ge 2$] always holds since the sending account balance and the sequencer's balance are always updated.

Reviewer: ```suggestion
* stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Notice that stem:[$m\ge 2$] is always true, because the balance of the account sending the transaction and the sequencer's balance are always updated.
```

* stem:[$t$] is the number of L2->L1 messages sent, where the corresponding payload sizes are denoted by stem:[$q_1,...,q_t$].
* stem:[$\ell$] is the number of contract deployments.
* stem:[$w$] is the xref:#calculation_of_computation_costs[`CairoResourceFeeWeights`] vector.

> * stem:[$\text{calldata_cost}$] is 550 gas per 32-byte word. This cost is derived as follows: 512 gas per 32-byte word for calldata, ~100gas for on-chain hashing that happens for every sent word, and 10% discount because the sequencer does not incur additional costs for repeated updates to the same storage slot within a single block, resulting in 0.9*612 = 550gas (rounded down).

Reviewer: ```suggestion
* stem:[$\text{calldata_cost}$] is 550 gas per 32-byte word. This cost is derived as follows: 512 gas per 32-byte word for calldata, ~100 gas for onchain hashing that happens for every word sent, and a 10% discount that applies because the sequencer does not incur additional costs for repeated updates to the same storage slot within a single block. So the gas cost is calculated as follows: stem:[0.9 \cot 612 = 550 gas], rounded down.
```

[stem]
++++
\underbrace{\textit{gas_price}\left(\text{calldata_cost} \cdot 2n - \text{contract_update_discount}\cdot n\right)}_{\text{contract addresses +  new nonce and number of storage updates
}} \\

+ \\

\underbrace{\textit{gas_price} \cdot \left(\text{calldata_cost}(2(m-2))+310\right)}_{\text{storage updates}}
++++

where:

* stem:[$n$] is the number of unique contracts updated.

> * stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Note that stem:[$m\ge 2$] always holds since the sending account balance and the sequencer's balance are always updated. The sender's balance is treated separately (described below) and the sequencer's balance is ignored since it is touched by all the transactions within a block.

Reviewer: ```suggestion
* stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Notice that stem:[$m\ge 2$] is always true,  because the balance of the account sending the transaction and the sequencer's balance are always updated. The sender's balance is treated separately, as described below, and the sequencer's balance is ignored since it is touched by all the transactions within a block.
```

> * stem:[\text{contract_update_discount}] is 312 gas, which is discounted for every updated contract, since out of the stem:[$2n$] words induced by updated contracts, stem:[$n$] are short, including at most stem:[$6$] non-zero bytes:

Reviewer: ```suggestion
* stem:[\text{contract_update_discount}] is 312 gas, which is discounted for every updated contract. This discount is a result of the fact that out of the stem:[$2n$] words caused by updating contracts, stem:[$n$] words are short, including at most 6 non-zero bytes:
```

+
--
** three bytes for the nonce
** two bytes for the number of storage updates
** one byte for the class information flag
--
+

> Taking into account that zero bytes only cost stem:[$4$] gas, the cost difference between a full 32-byte word (all none-zero) and a word with only six non-zero bytes is stem:[$32\cdot16-(6\cdot16+26\cdot4)=312$].

Reviewer: ```suggestion
Taking into account that zero bytes only cost 4 gas, the cost difference between a full 32-byte word, which is entirely non-zero, and a word with only six non-zero bytes is stem:[$32\cdot16-(6\cdot16+26\cdot4)=312$].
```

> * stem:[$310$] is the cost of updating the sender's balance, and is derived by assuming the balance requires at most 12 non-zero bytes (which is enough for 1.2B ETH or STRK), resulting in the following discounted price: stem:[$0.9*612-(512-(16*12+20*4))=310$] (rounded down).

Reviewer: ```suggestion
* stem:[$310$] is the cost of updating the sender's balance, and is derived by assuming the balance requires at most 12 non-zero bytes, which is enough for 1.2B ETH or STRK, resulting in the following discounted price: stem:[$0.9 \cdot 612-(512-(16 \cdot12+20 \cdot 4))=310$], rounded down.
```

[stem]
++++
\underbrace{\textit{gas_price}\left(\text{da_calldata_cost} \cdot 2n - \text{contract_update_discount}\cdot n\right)}_{\text{contract addresses +  new nonce and number of storage updates
}} \\

+ \\

\underbrace{\textit{gas_price} \cdot \left(\text{da_calldata_cost}(2(m-1))-240\right)}_{\text{storage updates}}
++++

where:

* stem:[$n$] is xref:#storage_updates[the number of unique contracts updated]. Note that this also takes into account contract deployment, even if the storage of the newly deployed contract was untouched, i.e. stem:[$n\ge\ell$].

> * stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Note that stem:[$m\ge 1$] always holds since the sequencer balance is always updated (this is not charged for).

Reviewer: ```suggestion
* stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Notice that stem:[$m\ge 1$] is always true, because the sequencer's balance is always updated, which does not incur any fee.
```

* stem:[\text{contract_update_discount}] is 312 gas, which is discounted for every updated contract. This discount is a result of the fact that out of the stem:[$2n$] words caused by updating contracts, stem:[$n$] words are short, including at most 6 non-zero bytes:
+
--
** three bytes for the nonce
** two bytes for the number of storage updates
** one byte for the class information flag
--
+
Taking into account that zero bytes only cost 4 gas, the cost difference between a full 32-byte word, which is entirely non-zero, and a word with only six non-zero bytes is stem:[$32\cdot16-(6\cdot16+26\cdot4)=312$].

> * stem:[$240$] is the gas discount for updating the sender's balance, and is derived by assuming the balance requires at most 12 non-zero bytes (which is enough for 1.2B ETH or STRK), resulting in the following discount: stem:[512-(20*4+12*16)=240$].

Reviewer: ```suggestion
* stem:[$240$] is the gas discount for updating the sender's balance, and is derived by assuming the balance requires at most 12 non-zero bytes, which is enough for 1.2B ETH or STRK, resulting in the following discount: stem:[512-(20\cdot4+12\cdot16)=240$].
```

[stem]
++++
\underbrace{\textit{gas_price}\left(\text{da_calldata_cost} \cdot 2n - \text{contract_update_discount}\cdot n\right)}_{\text{contract addresses +  new nonce and number of storage updates
}} \\

+ \\

\underbrace{\textit{gas_price} \cdot \left(\text{da_calldata_cost}(2(m-1))-240\right)}_{\text{storage updates}}
++++

where:

> * stem:[$n$] is xref:#storage_updates[the number of unique contracts updated]. Note that this also takes into account contract deployment, even if the storage of the newly deployed contract was untouched, i.e. stem:[$n\ge\ell$].

Reviewer: ```suggestion
* stem:[$n$] is xref:#storage_updates[the number of unique contracts updated], which includes changes to, or new deployments of, classes of existing contract instances, even if the storage of the newly deployed contract is untouched. In other words, stem:[$n\ge\ell$].
```

* stem:[$m$] is the number of values updated, not counting multiple updates for the same key. Notice that stem:[$m\ge 1$] is always true, because the sequencer's balance is always updated, which does not incur any fee.
* stem:[\text{contract_update_discount}] is 312 gas, which is discounted for every updated contract. This discount is a result of the fact that out of the stem:[$2n$] words caused by updating contracts, stem:[$n$] words are short, including at most 6 non-zero bytes:
+
--
** three bytes for the nonce
** two bytes for the number of storage updates
** one byte for the class information flag
--
+
Taking into account that zero bytes only cost 4 gas, the cost difference between a full 32-byte word, which is entirely non-zero, and a word with only six non-zero bytes is stem:[$32\cdot16-(6\cdot16+26\cdot4)=312$].

> * stem:[$240$] is the gas discount for updating the sender's balance, and is derived by assuming the balance requires at most 12 non-zero bytes, which is enough for 1.2B ETH or STRK, resulting in the following discount: stem:[512-(20\cdot4+12\cdot16)=240$].

Reviewer: ```suggestion
* stem:[$240$] is the gas discount for updating the sender's balance, and is derived by assuming the balance requires at most 12 non-zero bytes, which is enough for 1.2B ETH or STRK, resulting in the following discount: stem:[$512-(20\cdot4+12\cdot16)=240$].
```

* Payload size
* Payload (list of field elements)

> Consequently, the gas costs associated with a single L2→L1 message is:

Reviewer: ```suggestion
Consequently, the gas cost associated with a single L2→L1 message is:
```

[stem]
++++
\text{storage_write_cost} + \left(\text{da_calldata_cost} + \text{message_calldata_cost}\right)\cdot\left(3+\text{payload_size}\right)
++++

> Where stem:[$\text{da_calldata_cost}$] and stem:[$\text{message_calldata_cost}$] are 551 and 512 gas correspondingly.

Reviewer: ```suggestion
Where stem:[$\text{da_calldata_cost}$] is 551, and stem:[$\text{message_calldata_cost}$] is 512 gas.
```

> The reason messages appear twice in the fee formula (and pay both in stem:[$\text{da_calldata_cost}$] and stem:[$\text{message_calldata_cost}$]), 

Reviewer: ```suggestion
Messages appear twice in the fee formula, and pay both in stem:[$\text{da_calldata_cost}$] and stem:[$\text{message_calldata_cost}$]) because 
```

> is that they are sent to Ethereum twice: once to the STARK verifier contract (where the additional hashing cost is incured) and once when it is sent to the Starknet Core contract as part of the state update transaction.

Reviewer: ```suggestion
they are sent to Ethereum twice: once to the STARK verifier contract, where the additional hashing cost is incurred, and once when it is sent to the Starknet Core Contract as part of the state update transaction.
```

* number of storage updates and the new nonce
* class hash

> The first two elements are counted in the number of unique modified contracts which we denote by stem:[$n$] throughout this page (i.e. stem:[$n\ge\ell$]), so the only addition to the cost is coming from publishing the class hash,

Reviewer: ```suggestion
The first two elements are counted in the number of unique modified contracts, denoted by stem:[$n$] throughout this page, such as with stem:[$n\ge\ell$]. So the only additional comes from publishing the class hash,
```

> which adds 551 gas (see stem:[$\text{calldata_cost}$] above).

Reviewer: ```suggestion
which adds 551 gas. For more information, see stem:[$\text{calldata_cost}$] in xref:#fee_limitations[Fee limitations].
```

* number of storage updates and the new nonce
* class hash

The first two elements are counted in the number of unique modified contracts, denoted by stem:[$n$] throughout this page, such as with stem:[$n\ge\ell$]. So the only additional comes from publishing the class hash,

> which adds 551 gas. For more information, see stem:[$\text{calldata_cost}$] in xref:#fee_limitations[Fee limitations].

Reviewer: ```suggestion
which adds 551 gas. For more information, see stem:[$\text{da_calldata_cost}$] and stem:[$\text{message_calldata_cost}$] in xref:#l_2-l_1_messages[Onchain data: L2->L1 messages].
```
